File: apps/java/src/main/java/algorithm/BinarySearchTree.java
package algorithm;

import java.util.ArrayList;
import java.util.List;

/** 第9章 二分探索木 */
public class BinarySearchTree<T extends Comparable<T>> {

    public static class EmptyException extends RuntimeException {
        public EmptyException() { super("木は空です"); }
    }

    private static class BSTNode<T> {
        T key;
        BSTNode<T> left, right;
        BSTNode(T key) { this.key = key; }
    }

    private BSTNode<T> root;
    private int size;

    public BinarySearchTree() { root = null; size = 0; }

    public int size() { return size; }
    public boolean isEmpty() { return root == null; }
    public boolean contains(T key) { return search(key) != null; }

    private BSTNode<T> search(T key) {
        BSTNode<T> ptr = root;
        while (ptr != null) {
            int cmp = key.compareTo(ptr.key);
            if (cmp == 0) return ptr;
            ptr = cmp < 0 ? ptr.left : ptr.right;
        }
        return null;
    }

    public void insert(T key) {
        if (root == null) { root = new BSTNode<>(key); size++; return; }
        BSTNode<T> ptr = root;
        while (true) {
            int cmp = key.compareTo(ptr.key);
            if (cmp == 0) return; // 重複は無視
            if (cmp < 0) {
                if (ptr.left == null) { ptr.left = new BSTNode<>(key); size++; return; }
                ptr = ptr.left;
            } else {
                if (ptr.right == null) { ptr.right = new BSTNode<>(key); size++; return; }
                ptr = ptr.right;
            }
        }
    }

    public void delete(T key) {
        BSTNode<T> parent = null;
        BSTNode<T> ptr = root;
        boolean isLeftChild = false;

        while (ptr != null) {
            int cmp = key.compareTo(ptr.key);
            if (cmp == 0) break;
            parent = ptr;
            if (cmp < 0) { isLeftChild = true; ptr = ptr.left; }
            else { isLeftChild = false; ptr = ptr.right; }
        }
        if (ptr == null) return;

        size--;

        if (ptr.left == null && ptr.right == null) {
            replaceNode(parent, isLeftChild, null);
        } else if (ptr.right == null) {
            replaceNode(parent, isLeftChild, ptr.left);
        } else if (ptr.left == null) {
            replaceNode(parent, isLeftChild, ptr.right);
        } else {
            // 右部分木の最小ノード（中順後継）で置き換え
            BSTNode<T> succParent = ptr;
            BSTNode<T> succ = ptr.right;
            while (succ.left != null) { succParent = succ; succ = succ.left; }
            ptr.key = succ.key;
            if (succParent == ptr) succParent.right = succ.right;
            else succParent.left = succ.right;
            size++; // delete で既に -1 したので +1 で調整
        }
    }

    private void replaceNode(BSTNode<T> parent, boolean isLeftChild, BSTNode<T> newNode) {
        if (parent == null) root = newNode;
        else if (isLeftChild) parent.left = newNode;
        else parent.right = newNode;
    }

    public T min() {
        if (root == null) throw new EmptyException();
        BSTNode<T> ptr = root;
        while (ptr.left != null) ptr = ptr.left;
        return ptr.key;
    }

    public T max() {
        if (root == null) throw new EmptyException();
        BSTNode<T> ptr = root;
        while (ptr.right != null) ptr = ptr.right;
        return ptr.key;
    }

    public List<T> inOrder() {
        List<T> result = new ArrayList<>();
        inOrderHelper(root, result);
        return result;
    }

    private void inOrderHelper(BSTNode<T> node, List<T> result) {
        if (node == null) return;
        inOrderHelper(node.left, result);
        result.add(node.key);
        inOrderHelper(node.right, result);
    }

    public List<T> preOrder() {
        List<T> result = new ArrayList<>();
        preOrderHelper(root, result);
        return result;
    }

    private void preOrderHelper(BSTNode<T> node, List<T> result) {
        if (node == null) return;
        result.add(node.key);
        preOrderHelper(node.left, result);
        preOrderHelper(node.right, result);
    }

    public List<T> postOrder() {
        List<T> result = new ArrayList<>();
        postOrderHelper(root, result);
        return result;
    }

    private void postOrderHelper(BSTNode<T> node, List<T> result) {
        if (node == null) return;
        postOrderHelper(node.left, result);
        postOrderHelper(node.right, result);
        result.add(node.key);
    }
}

File: apps/csharp/Algorithm/BinarySearchTree.cs
namespace Algorithm;

/// <summary>第9章 二分探索木</summary>
public class BinarySearchTree<T> where T : IComparable<T>
{
    public class EmptyException() : Exception("木は空です");

    private class BSTNode(T key) { public T Key = key; public BSTNode? Left, Right; }

    private BSTNode? _root;
    private int _size;

    public int Size() => _size;
    public bool IsEmpty() => _root == null;
    public bool Contains(T key) => Search(key) != null;

    private BSTNode? Search(T key)
    {
        var ptr = _root;
        while (ptr != null)
        {
            int cmp = key.CompareTo(ptr.Key);
            if (cmp == 0) return ptr;
            ptr = cmp < 0 ? ptr.Left : ptr.Right;
        }
        return null;
    }

    public void Insert(T key)
    {
        if (_root == null) { _root = new BSTNode(key); _size++; return; }
        var ptr = _root;
        while (true)
        {
            int cmp = key.CompareTo(ptr.Key);
            if (cmp == 0) return;
            if (cmp < 0) { if (ptr.Left == null) { ptr.Left = new BSTNode(key); _size++; return; } ptr = ptr.Left; }
            else { if (ptr.Right == null) { ptr.Right = new BSTNode(key); _size++; return; } ptr = ptr.Right; }
        }
    }

    public void Delete(T key)
    {
        BSTNode? parent = null; var ptr = _root; bool isLeft = false;
        while (ptr != null)
        {
            int cmp = key.CompareTo(ptr.Key);
            if (cmp == 0) break;
            parent = ptr;
            if (cmp < 0) { isLeft = true; ptr = ptr.Left; } else { isLeft = false; ptr = ptr.Right; }
        }
        if (ptr == null) return;
        _size--;
        if (ptr.Left == null && ptr.Right == null) Replace(parent, isLeft, null);
        else if (ptr.Right == null) Replace(parent, isLeft, ptr.Left);
        else if (ptr.Left == null) Replace(parent, isLeft, ptr.Right);
        else
        {
            var succParent = ptr; var succ = ptr.Right;
            while (succ.Left != null) { succParent = succ; succ = succ.Left; }
            ptr.Key = succ.Key;
            if (succParent == ptr) succParent.Right = succ.Right; else succParent.Left = succ.Right;
            _size++;
        }
    }

    private void Replace(BSTNode? parent, bool isLeft, BSTNode? node)
    {
        if (parent == null) _root = node;
        else if (isLeft) parent.Left = node;
        else parent.Right = node;
    }

    public T Min() { if (_root == null) throw new EmptyException(); var p = _root; while (p.Left != null) p = p.Left; return p.Key; }
    public T Max() { if (_root == null) throw new EmptyException(); var p = _root; while (p.Right != null) p = p.Right; return p.Key; }

    public List<T> InOrder() { var r = new List<T>(); InOrderHelper(_root, r); return r; }
    private void InOrderHelper(BSTNode? n, List<T> r) { if (n == null) return; InOrderHelper(n.Left, r); r.Add(n.Key); InOrderHelper(n.Right, r); }

    public List<T> PreOrder() { var r = new List<T>(); PreOrderHelper(_root, r); return r; }
    private void PreOrderHelper(BSTNode? n, List<T> r) { if (n == null) return; r.Add(n.Key); PreOrderHelper(n.Left, r); PreOrderHelper(n.Right, r); }

    public List<T> PostOrder() { var r = new List<T>(); PostOrderHelper(_root, r); return r; }
    private void PostOrderHelper(BSTNode? n, List<T> r) { if (n == null) return; PostOrderHelper(n.Left, r); PostOrderHelper(n.Right, r); r.Add(n.Key); }
}
